xonsh/environ.py

```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
                 p.break_()
                 p.pretty(dict(self))
 
+    def register(self, varname, vartype=None, defaultval=None, vardoc=None,
+                 ens_validate=None, ens_convert=None, ens_detype=None,
+                 doc_configurable=None, doc_default=None, doc_store_as_str=None):
+        """Register an enviornment variable with optional type handling,
+        default value, doc.
+
+        Parameters
+        ----------
+        varname : str
+            Environment variable name to register. Typically all caps.
+        vartype : str, optional,  {'bool', 'str', 'path', 'int', 'float'}
+            Variable type. If not one of the available presets, use `ens_validate`,
+            `ens_convert`, and `ens_detype` to specify type behavior.
+        defaultval : optional
+            Default value for variable. ``ValueError`` raised if type does not match
+            that specified by `vartype` (or `ens_validate`).
+        vardoc : str, optional
+            Docstring for variable.
+        ens_validate : func, optional
+            Function to validate type.
+        ens_convert : func, optional
+            Function to convert variable from a string representation to its type.
+        ens_detype : func, optional
+            Function to convert variable from its type to a string representation.
+        doc_configurable : bool, optional
+            Flag for whether the environment variable is configurable or not.
+        doc_default : str, optional
+            Custom docstring for the default value for complex defaults.
+            If this is ``DefaultNotGiven``, then the default will be looked up
+            from ``DEFAULT_VALUES`` and converted to a ``str``.
+        doc_store_as_str : bool, optional
+            Flag for whether the environment variable should be stored as a
+            string. This is used when persisting a variable that is not JSON
+            serializable to the config file. For example, sets, frozensets, and
+            potentially other non-trivial data types. default, False.
+
+        """
+
+        if ((vartype is not None) and 
+            (vartype in ('bool', 'str', 'path', 'int', 'float'))):
+            ensurer = Ensurer(*ENSURERS[vartype])
+        else:
+            ensurer = Ensurer(*(ens_validate, ens_convert, ens_detype))
+
+        # set ensurer for envvar
+        set_ensurer(varname, ensurer)
+
+        # set default value for envvar
+        if defaultval is not None:
+            if ensurer['validate'](defaultval):
+                self._defaults[varname] = defaultval
+            else:
+                raise ValueError("Default value does not match type specified by ensurer")
+
+        # set doc for envvar
+        if vardoc is not None:
+            self._docs[varname] = VarDocs(
```

Comment:
Ideally I think we'd combine VarDocs and Ensurer into a single namedtuple

Reply:
Can do. Do we want to merge `DEFAULT_ENSURERS` and `DEFAULT_DOCS` along with this change (seems inevitable). Would we also want to merge in `DEFAULT_VALUES` to this new namedtuple?

Reply:
ping @scopatz; do we want to merge together all the `DEFAULT_*` lazy objects?

Reply:
Yes, I think we do.  I am so sorry for taking so long to get back to this here

Reply:
No worries, all good! I will proceed with implementation. I believe I can carve out some time next week.

Reply:
Awesome! thanks @dotsdl!